sessionization/sessionization.md

```diff
@@ -0,0 +1,597 @@
+        }
+      } else {
+        break; // stop, later windows won't intersect either
+      }
+    }
+    if (first != null && last != null) {
+      // remove entire subset from from first to last after it got merged into acc
+      buff.navigableKeySet().subSet(first, true, last, true).clear();
+    }
+
+    if (acc == null) {
+      // init accumulator if necessary
+      acc = fn.createAccumulator();
+    }
+    // add input and set accumulator for new (potentially merged) window
+    buff.put(window, fn.addInput(acc, input.getValue().getValue()));
+  }
+  return buff;
+}
+```
+
+
+##### Merging aggregation buffers
+
+The idea to merge our tree buffers is simple; we traverse both trees in order and always pick the next smallest window (in time). While it intersects a previous one, we expan the windows and merge the accumulators, if not we add it to a new tree buffer and continue until there's no more windows; once the final window is added we're done.
+
+The following code implements the merge operation. Once again, the full, feature-complete implementation is available on [<img src="github.png" style="height:16px;"/> apache/beam](https://github.com/apache/beam/blob/3a4d57eb8976c5f503b32d478a80b1800490f66f/runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/Aggregators.java#L231-L271).
+
+```java
+public TreeMap<IntervalWindow, AccumT> merge(TreeMap<IntervalWindow, AccumT> buff1, TreeMap<IntervalWindow, AccumT> buff2) {
+  if (buff1.isEmpty()) {
+    return buff2;
+  } else if (buff2.isEmpty()) {
+    return buff1;
+  }
+  // Init new tree map to merge both buffers
+  TreeMap<IntervalWindow, AccumT> res = zero();
+  PeekingIterator<Map.Entry<IntervalWindow, AccumT>> it1 =
+      Iterators.peekingIterator(buff1.entrySet().iterator());
+  PeekingIterator<Map.Entry<IntervalWindow, AccumT>> it2 =
+      Iterators.peekingIterator(buff2.entrySet().iterator());
+
+  AccumT acc = null;
+  IntervalWindow window = null;
+  while (it1.hasNext() || it2.hasNext()) {
+    // pick iterator with the smallest window ahead and forward it
+    Map.Entry<IntervalWindow, AccumT> nextMin =
+        (it1.hasNext() && it2.hasNext())
+            ? it1.peek().getKey().compareTo(it2.peek().getKey()) <= 0 ? it1.next() : it2.next()
+            : it1.hasNext() ? it1.next() : it2.next();
+    if (window != null && window.intersects(nextMin.getKey())) {
+      // extend window and merge accumulators if windows intersect
+      window = window.span(nextMin.getKey());
+      acc = fn.mergeAccumulators(ImmutableList.of(acc, nextMin.getValue()));
+    } else {
+      // store window / accumulator if necessary and continue with next minimum
+      if (window != null && acc != null) {
+        res.put(window, acc);
+      }
+      acc = nextMin.getValue();
+      window = nextMin.getKey();
+    }
+  }
+  if (window != null && acc != null) {
+    res.put(window, acc);
+  }
+  return res;
+}
+```
+
+## Benchmarks
```

Comment:
Not required but adding TPC-DS results also would be helpful

Reply:
@aromanenko-dev I just had a look and don't think this applies to the TPC-DS benchmarks. I couldn't find any queries using session windows, but also don't know how that would be expressed in SQL. Let me know if that's not the case

Reply:
Right, agree